File: ideaproject/javaserializer/src/com/houseelectrics/serializer/PropertyReflectionNodeExpander.java
package com.houseelectrics.serializer;

import java.lang.reflect.*;

public class PropertyReflectionNodeExpander implements NodeExpander
{
    private boolean excludeReadOnlyProperties = false;
    public boolean getExcludeReadOnlyProperties() { return excludeReadOnlyProperties; }
    public void setExcludeReadOnlyProperties(boolean value) { this.excludeReadOnlyProperties = value; }
    public int expand(Object o, NodeExpansionListener listener)
    {
        Method[]ms= o.getClass().getMethods();
        int setAbleCount = 0;

        for  (Method m : ms)
        {
            if (!Modifier.isPublic(m.getModifiers())) continue;
            if (!m.getName().startsWith("get")) continue;
            if (m.getDeclaringClass()==Object.class) continue;
            Class []prms = m.getParameterTypes();
            if (prms!=null && prms.length>0) continue;
            if (excludeReadOnlyProperties)
            {
                String setMethodName = "set" + m.getName().substring(3);
                Class []paramTypes = {m.getReturnType()};
                Method setMethod = null;
                try {
                    setMethod = o.getClass().getMethod(setMethodName, paramTypes);
                }
                catch (NoSuchMethodException ex)
                {
                    //todo - eliminate this !
                }
                if (setMethod==null) {continue;}
            }
            Object value = null;
            try
            {
                Object []params={};
                value = m.invoke(o, params);
                setAbleCount++;
            }
            catch (Exception ex)
            {
                RuntimeException exx = new RuntimeException("failed to get value for property " + m.getName() + " declared in "+ m.getDeclaringClass().getName(), ex);
                throw exx;
            }

            String name =   m.getName().substring(3);//Character.toLowerCase(m.getName().charAt(3) )  +   m.getName().substring(4);
            listener.OnChildNode(o, name, value);
        }
        return setAbleCount;
    }
}

File: ideaproject/javaserializer/src/com/houseelectrics/serializer/PropertyReflectionNodeExpander.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;

namespace com.houseelectrics.serializer
{
    public class PropertyReflectionNodeExpander : NodeExpander
    {
        private bool excludeReadOnlyProperties = false;
        public bool ExcludeReadOnlyProperties { get { return excludeReadOnlyProperties; } set { this.excludeReadOnlyProperties = value; } }
        public int expand(object o, OnChildNode onChildNode)
        {
            PropertyInfo []pis= o.GetType().GetProperties();
            int setAbleCount = 0;
            foreach (PropertyInfo pi in pis)
            {
                if (ExcludeReadOnlyProperties && pi.GetSetMethod() == null) { continue; }
                Object value = null;
                try
                {
                    value = pi.GetValue(o);
                    setAbleCount++;
                }
                catch (Exception ex)
                {
                    Exception exx = new Exception("failed to get value for property " + pi.Name + " declared in "+ pi.DeclaringType, ex);
                    throw exx;
                }
                
                String name = Char.ToLower(pi.Name[0]) + pi.Name.Substring(1);
                name = pi.Name;
                onChildNode(o, name, value);
            }
            return setAbleCount;
        }
    }
}
